### ISSUE TITLE
Use the NodeJS >= v18.2.0 closeAllConnections and closeIdleConnections

### ISSUE THREAD

**Issue by mstaicu** [LuKks/graceful-http#3]
Really nice terse utility for handling graceful HTTP server shutdowns. Was wondering if there are plans on the roadmap to switch to using the _closeAllConnections_ and _closeIdleConnections_ methods. I assume they will change the implementation logic of this utility.

**Comment by LuKks** [comment]
Yes 100%, see related https://github.com/LuKks/graceful-http/pull/2#issuecomment-2000244036

After some thinking, this lib could still provide value like automatically handling the Connection header, force close timeout, etc

We should probably wait for another Node.js LTS so we make sure it's compatible with most current systems, this avoids conditionals

**Comment by mstaicu** [comment]
> Yes 100%, see related [#2 (comment)](https://github.com/LuKks/graceful-http/pull/2#issuecomment-2000244036)
> 
> After some thinking, this lib could still provide value like automatically handling the Connection header, force close timeout, etc
> 
> We should probably wait for another Node.js LTS so we make sure it's compatible with most current systems, this avoids conditionals

Definitely +1 on the value, especially since you can use this with clustering on worker processes. So we should wait for the next LTS then revisit it, correct?

**Comment by mstaicu** [comment]
Also, off topic but, would this lib be a good fit for wrapping an express.js web app running in a Kubernetes pod? I assume it is, since Kubernetes would be responsible for sending SIGTERM and SIGINIT and perform health checks. I also have a Traefik instance in front of all my services which does SSL termination and HTTPS to HTTP downgrade for internal comms

**Comment by LuKks** [comment]
As for clustering, yes I'm currently using it everywhere, even in the reverse proxy!

Router: https://github.com/LuKks/simple-hosting/blob/72271544c38b2eb9e15edf059cab649450fc94d7/index.js#L29
Backends: https://github.com/like/like-backend/blob/8211e9cdc2e78579bd3293c052569b4073fc40bb/index.js#L24

My current router is a `simple-hosting` app that acts as reverse proxy which it has `graceful-http` built-in, and a separated cluster script where when I want to reload it first spawns a new worker then it gracefully stops the older one. Probably Traefik already takes care of this

My current backends are not using cluster yet but planning to reuse that cluster script for everything, since `like-backend` also have `graceful-http` built-in then it just works haha. I use systemd for spawning Node processes, so if Kubernetes sends the signal as systemd would then it's all good (I have to use containers like you soon, considering systemd-spawn or Docker atm)

If you're interested in the cluster script I will see to make it public sooner, it's a very intense quickly done script but it works

Also interested on opinions or ideas to offer public services on how to improve those setups in general, classic old repetitive problem :p

**Comment by mstaicu** [comment]
Don't want to start a new issue to propose something so I'll write it up here. I was thinking of adding a new WeakMap to track all responses on a per socket basis, something like `WeakMap<Socket, ServerResponse>`, instead of the current responses map. Not knowing the HTTP/1.1 protocol in and out but I'm guessing that you could have multiple requests on a per socket basis, and only once all those requests have been added the `Connection: close` header we could consider the socket a valid candidate for closing it. Again, not sure if this logic adds any benefits, just running it by you to get some feedback on it
